File: vfr-green-app/app/src/main/java/com/blueskycharts/app/map/utility/PriorityArray.kt
package com.blueskycharts.app.map.utility

import java.util.*
import kotlin.collections.ArrayList

class PriorityArray<T> {
    private var content: ArrayList<LinkedList<T>> = ArrayList()
    private var size: Int = 0
    private var lowestPriority: Int = 0

    fun clear() {
        this.content = ArrayList();
        this.size = 0;
    }

    fun add(obj: T, priority: Int) {
        // Setup the priority and set the size of this object
        if ( this.size == 0 || this.lowestPriority > priority ) {
            this.lowestPriority = priority;
        }
        this.size++;
        while ( this.content.size <= priority ) {
            this.content.add(LinkedList<T>());
        }

        // Insert the object
        this.content[priority].add(obj);
    }

    fun pop(): T? {
        if ( this.size == 0 ) {
            return null;
        }
        val retVal = this.content[this.lowestPriority].removeFirst() ?: return null;

        // Adjust the lowest priority if we've used this one up
        while ( this.content[this.lowestPriority].size == 0 ) {
            this.lowestPriority++;
        }

        this.size--;
        return retVal;
    }

    fun size(): Int {
        return this.size;
    }

    fun getPriorityList(priority: Int): Collection<T> {
        return this.content[priority];
    }
}

File: vfr-green-map/src/utility/priority-array.ts

export class PriorityArray<T> {
    private content: Array<Array<T>>;
    private s: number;
    private lowestPriority: number;

    constructor() {
        this.content = [[]];
        this.s = 0;
        this.lowestPriority = 0;
    }

    public clear(): void {
        this.content = [];
        this.s = 0;
    }

    public add(obj: T, priority: number): void {
        // Setup the priority and set the size of this object
        if ( this.s == 0 || this.lowestPriority > priority ) {
            this.lowestPriority = priority;
        }
        this.s++;
        while ( this.content.length <= priority ) {
            this.content.push([]);
        }

        // Insert the object
        this.content[priority].push(obj);
    }

    public pop(): T {
        this.s--;
        let retVal = this.content[this.lowestPriority].pop();

        // Adjust the lowest priority if we've used this one up
        while ( this.content[this.lowestPriority].length == 0 && this.s > 0 ) {
            this.lowestPriority++;
        }

        // Return the object
        if ( retVal === undefined ) {
            return this.pop();
        }
        return retVal;
    }

    public size(): number {
        return this.s;
    }

    public getPriorityList(priority: number): Array<T> {
        return this.content[priority];
    }
}